en/lessons/specifics/benchmarking.md

```diff
@@ -0,0 +1,346 @@
+defp deps do
+  [{:benchee, "~> 0.9", only: :dev}]
+end
+```
+Then we call:
+
+```shell
+$ mix deps.get
+...
+$ mix deps.compile
+```
+
+The first command will download and install Benchee. You may be asked to install Hex along with it. The second compiles the Benchee application. Now we're ready to write our first benchmark!
+
+**An important note before we begin:** When benchmarking, it is very important to not use `iex` since that behaves differently and is often much slower than how your code is most likely being used in production. So, let's create a file that we're calling `benchmark.exs`, and in that file we'll add the following code:
+
+```elixir
+list = Enum.to_list(1..10_000)
+map_fun = fn(i) -> [i, i * i] end
+
+Benchee.run(%{
+  "flat_map"    => fn -> Enum.flat_map(list, map_fun) end,
+  "map.flatten" => fn -> list |> Enum.map(map_fun) |> List.flatten end
+})
+```
+
+Now to run our benchmark, we call:
+
+```shell
+$ mix run benchmark.exs
+```
+
+And we should see something like the following output in your console:
+
+```shell 
+Operating System: macOS
+CPU Information: Intel(R) Core(TM) i5-4260U CPU @ 1.40GHz
+Number of Available Cores: 4
+Available memory: 8.589934592 GB
+Elixir 1.5.1
+Erlang 20.0
+Benchmark suite executing with the following configuration:
+warmup: 2.00 s
+time: 5.00 s
+parallel: 1
+inputs: none specified
+Estimated total run time: 14.00 s
+
+
+Benchmarking flat_map...
+Benchmarking map.flatten...
+
+Name                  ips        average  deviation         median
+flat_map           1.03 K        0.97 ms    ±33.00%        0.85 ms
+map.flatten        0.56 K        1.80 ms    ±31.26%        1.60 ms
+
+Comparison:
+flat_map           1.03 K
+map.flatten        0.56 K - 1.85x slower
+```
+
+Of course your system information and results may be different depending on the specifications of the machine you are running your benchmarks on, but this general information should all be there.
+
+At first glance, the `Comparison` section shows us that our `map.flatten` version is 1.85x slower than `flat_map` - very helpful to know! But let's look at the other statistics that we get:
+
+* **ips** - this stands for "iterations per second," which tells us how often the given function can be executed in one second. For this metric, a higher number is better.
+* **average** - this is the average execution time of the given function. For this metric, a lower number is better.
+* **deviation** - this is the standard deviation, which tells us how much the results for each iteration varies in the results. Here it is given as a percentage of the average.
+* **median** - when all measured times are sorted, this is the middle value (or average of the two middle values when the number of samples is even). Because of some BEAM settings that we're unable to modify, this will be more stable than the `average`, and somewhat more likely to reflect the normal performance of your code in production. For this metric, a lower number is better.
```

Comment:
not only BEAM settings but just overall system load, might just keep it general and say that some iterations may be signigificantly slower due to the environment